File: packages/fsm-core-async-op-worker/src/worker-sdk/python/validate_async_operation.py
"""Discovers and verifies python promise-actor folders.

Python counterpart of
../typescript/validate-async-operation.ts's
validateAsyncOperationFromFoldersTypescript, walking
<folder_path>/<fsm_name>/<version>/python/actors/<actor_name>/<actor_name>.py
(this repo's real FSM actor convention — see
apps/fsm-core-example/fsm/creditCheck/v01/python/actors/checkBureau/checkBureau.py)
instead of the typescript file's .../typescript/actors/... path.

Verification inlines packages/fsm-compiler-ts/src/checkers/check_fn.py's own
check (ast.walk for a FunctionDef/AsyncFunctionDef matching the actor name)
directly in-process, the same way the typescript file inlines check_fn.ts's
check instead of shelling out to it — this file only ever checks python, in
a process that's already Python.
"""

from __future__ import annotations

import ast
import os
import re
from dataclasses import dataclass
from typing import List, Optional

LANG = "python"

_VERSION_FOLDER_RE = re.compile(r"^v\d{2}$")


def _is_version_folder_name(name: str) -> bool:
    # Mirrors @pgfsm/compiler's util.ts isVersionFolderName: /^v\d{2}$/.
    return bool(_VERSION_FOLDER_RE.match(name))


def _has_fn(source: str, fn_name: str) -> bool:
    for node in ast.walk(ast.parse(source)):
        if (
            isinstance(node, (ast.FunctionDef, ast.AsyncFunctionDef))
            and node.name == fn_name
        ):
            return True
    return False


@dataclass
class ActorPluginValidationResult:
    """Mirrors @pgfsm/compiler's ActorPluginValidationResult (util.ts)."""

    src: str
    method: str
    fsm_name: str
    fsm_type: str
    fsm_version: str
    fsm_language: str
    is_verified: bool
    fsm_module_path: str
    parent_fsm_name: str
    parent_fsm_version: str
    comment: str
    parent_fsm_path: str
    error_message: Optional[str]


def validate_async_operation_from_folders_python(
    folder_path: str,
    workflow_type: str,
    skip_dirs: Optional[List[str]] = None,
) -> List[ActorPluginValidationResult]:
    del workflow_type  # unused — kept for parity with validateAsyncOperationFromFolders
    skip_dirs = skip_dirs or []

    if folder_path.startswith("."):
        raise ValueError(
            f"Invalid folder path: {folder_path}. Folder paths cannot start with '.'"
        )
    if folder_path.endswith("/"):
        raise ValueError(
            f"Invalid folder path: {folder_path}. Folder paths cannot end with '/'"
        )

    abs_folder_path = (
        folder_path
        if folder_path.startswith("/")
        else os.path.join(os.getcwd(), folder_path)
    )

    results: List[ActorPluginValidationResult] = []

    if not os.path.isdir(abs_folder_path):
        print(f"Provided path is not a directory: {abs_folder_path}")
        return results

    for fsm_name in sorted(os.listdir(abs_folder_path)):
        if fsm_name in skip_dirs:
            continue
        fsm_dir_path = os.path.join(abs_folder_path, fsm_name)
        if not os.path.isdir(fsm_dir_path):
            continue

        for version in sorted(os.listdir(fsm_dir_path)):
            version_path = os.path.join(fsm_dir_path, version)
            if not os.path.isdir(version_path):
                continue
            if not _is_version_folder_name(version):
                continue

            lang_path = os.path.join(version_path, LANG)
            if not os.path.isdir(lang_path):
                continue

            actors_path = os.path.join(lang_path, "actors")
            if not os.path.isdir(actors_path):
                continue

            for actor_name in sorted(os.listdir(actors_path)):
                actor_dir = os.path.join(actors_path, actor_name)
                if not os.path.isdir(actor_dir):
                    continue

                module_path = os.path.join(actor_dir, f"{actor_name}.py")
                if not os.path.isfile(module_path):
                    continue

                is_verified = False
                error_message: Optional[str] = None
                try:
                    with open(module_path) as f:
                        source = f.read()
                    if _has_fn(source, actor_name):
                        is_verified = True
                    else:
                        error_message = (
                            f"Function '{actor_name}' not found in {module_path}"
                        )
                except (OSError, SyntaxError) as exc:
                    error_message = f"Failed to parse {module_path}: {exc}"

                results.append(
                    ActorPluginValidationResult(
                        src=actor_name,
                        method=actor_name,
                        fsm_name=actor_name,
                        fsm_type="promise",
                        fsm_version=version,
                        fsm_language=LANG,
                        is_verified=is_verified,
                        fsm_module_path=module_path,
                        parent_fsm_name=fsm_name,
                        parent_fsm_version=version,
                        comment="fsmVersion is parentFsmVersion for fsmType promise",
                        parent_fsm_path=fsm_dir_path,
                        error_message=error_message,
                    )
                )

    return results

File: packages/fsm-core-async-op-worker/src/worker-sdk/rust/src/validate_async_operation.rs
//! Discovers and verifies rust promise-actor folders.
//!
//! Rust counterpart of ../typescript/validate-async-operation.ts's
//! validateAsyncOperationFromFoldersTypescript, walking
//! <folder_path>/<fsm_name>/<version>/rust/actors/<actor_name>/<actor_name>.rs
//! (this repo's real FSM actor convention — see
//! apps/fsm-core-example/fsm/creditCheck/v01/rust/actors/checkBureau/checkBureau.rs)
//! instead of the typescript file's .../typescript/actors/... path.
//!
//! Verification inlines packages/fsm-compiler-ts/src/checkers/check_fn.rs's
//! own check (substring match for `pub fn <name>(` / `pub async fn
//! <name>(`) directly in-process, the same way the typescript and python
//! files inline their own checkers instead of shelling out — this file
//! only ever checks rust, in a process that's already... not Rust source
//! being compiled per actor, just plain text pattern matching, so no
//! subprocess/compiler invocation is needed here either.
//!
//! This is validation only, not invocation: unlike Python/TypeScript, a
//! verified result here does not carry a callable — Rust has no runtime
//! mechanism to load a function out of a `.rs` file (see sdk.rs's module
//! doc). The reference binary (main.rs) still has to match verified
//! results against a small compile-time registry of actually-linked-in
//! handler functions.

use std::fs;
use std::path::Path;

const LANG: &str = "rust";

fn is_version_folder_name(name: &str) -> bool {
    // Mirrors @pgfsm/compiler's util.ts isVersionFolderName: /^v\d{2}$/.
    let bytes = name.as_bytes();
    bytes.len() == 3 && bytes[0] == b'v' && bytes[1].is_ascii_digit() && bytes[2].is_ascii_digit()
}

fn has_fn(source: &str, fn_name: &str) -> bool {
    let patterns = [
        format!("pub fn {}(", fn_name),
        format!("pub fn {} (", fn_name),
        format!("pub async fn {}(", fn_name),
        format!("pub async fn {} (", fn_name),
    ];
    patterns.iter().any(|pat| source.contains(pat.as_str()))
}

/// Mirrors @pgfsm/compiler's ActorPluginValidationResult (util.ts) field for
/// field, for parity with the TS/Python versions — main.rs only reads a
/// subset (fsm_name, fsm_type, fsm_version, parent_fsm_*, is_verified,
/// fsm_module_path, error_message); the rest exist for completeness/future
/// consumers.
#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct ActorPluginValidationResult {
    pub src: String,
    pub method: String,
    pub fsm_name: String,
    pub fsm_type: String,
    pub fsm_version: String,
    pub fsm_language: String,
    pub is_verified: bool,
    pub fsm_module_path: String,
    pub parent_fsm_name: String,
    pub parent_fsm_version: String,
    pub parent_fsm_path: String,
    pub error_message: Option<String>,
}

pub fn validate_async_operation_from_folders_rust(
    folder_path: &str,
    skip_dirs: &[String],
) -> Vec<ActorPluginValidationResult> {
    let mut results = Vec::new();

    let abs_folder_path = Path::new(folder_path);
    if !abs_folder_path.is_dir() {
        eprintln!("Provided path is not a directory: {}", folder_path);
        return results;
    }

    let Ok(fsm_entries) = fs::read_dir(abs_folder_path) else {
        return results;
    };
    let mut fsm_dirs: Vec<_> = fsm_entries.flatten().collect();
    fsm_dirs.sort_by_key(|e| e.file_name());

    for fsm_entry in fsm_dirs {
        let fsm_name = fsm_entry.file_name().to_string_lossy().to_string();
        if skip_dirs.iter().any(|d| d == &fsm_name) {
            continue;
        }
        let fsm_dir_path = fsm_entry.path();
        if !fsm_dir_path.is_dir() {
            continue;
        }

        let Ok(version_entries) = fs::read_dir(&fsm_dir_path) else {
            continue;
        };
        let mut version_dirs: Vec<_> = version_entries.flatten().collect();
        version_dirs.sort_by_key(|e| e.file_name());

        for version_entry in version_dirs {
            let version = version_entry.file_name().to_string_lossy().to_string();
            let version_path = version_entry.path();
            if !version_path.is_dir() || !is_version_folder_name(&version) {
                continue;
            }

            let lang_path = version_path.join(LANG);
            if !lang_path.is_dir() {
                continue;
            }

            let actors_path = lang_path.join("actors");
            if !actors_path.is_dir() {
                continue;
            }

            let Ok(actor_entries) = fs::read_dir(&actors_path) else {
                continue;
            };
            let mut actor_dirs: Vec<_> = actor_entries.flatten().collect();
            actor_dirs.sort_by_key(|e| e.file_name());

            for actor_entry in actor_dirs {
                let actor_name = actor_entry.file_name().to_string_lossy().to_string();
                let actor_dir = actor_entry.path();
                if !actor_dir.is_dir() {
                    continue;
                }

                let module_path = actor_dir.join(format!("{}.rs", actor_name));
                if !module_path.is_file() {
                    continue;
                }

                let (is_verified, error_message) = match fs::read_to_string(&module_path) {
                    Ok(source) if has_fn(&source, &actor_name) => (true, None),
                    Ok(_) => (
                        false,
                        Some(format!(
                            "Function '{}' not found in {}",
                            actor_name,
                            module_path.display()
                        )),
                    ),
                    Err(err) => (
                        false,
                        Some(format!("Failed to read {}: {}", module_path.display(), err)),
                    ),
                };

                results.push(ActorPluginValidationResult {
                    src: actor_name.clone(),
                    method: actor_name.clone(),
                    fsm_name: actor_name,
                    fsm_type: "promise".to_string(),
                    fsm_version: version.clone(),
                    fsm_language: LANG.to_string(),
                    is_verified,
                    fsm_module_path: module_path.display().to_string(),
                    parent_fsm_name: fsm_name.clone(),
                    parent_fsm_version: version.clone(),
                    parent_fsm_path: fsm_dir_path.display().to_string(),
                    error_message,
                });
            }
        }
    }

    results
}
